rpc-rs/src/otel.rs

```diff
@@ -18,29 +18,41 @@ lazy_static::lazy_static! {
 #[derive(Debug)]
 pub struct OtelHeaderExtractor<'a> {
     inner: &'a HeaderMap,
+    keys: Vec<String>,
 }
 
 impl<'a> OtelHeaderExtractor<'a> {
     /// Creates a new extractor using the given [`HeaderMap`]
     pub fn new(headers: &'a HeaderMap) -> Self {
-        OtelHeaderExtractor { inner: headers }
+        OtelHeaderExtractor {
+            inner: headers,
+            keys: headers
+                .iter()
+                .map(|(k, _)| String::from_utf8_lossy(k.as_ref()).to_string())
+                .collect(),
+        }
     }
 
     /// Creates a new extractor using the given message
     pub fn new_from_message(msg: &'a async_nats::Message) -> Self {
+        let inner = msg.headers.as_ref().unwrap_or(&EMPTY_HEADERS);
         OtelHeaderExtractor {
-            inner: msg.headers.as_ref().unwrap_or(&EMPTY_HEADERS),
+            inner,
+            keys: inner
+                .iter()
+                .map(|(k, _)| String::from_utf8_lossy(k.as_ref()).to_string())
+                .collect(),
         }
     }
 }
 
 impl<'a> Extractor for OtelHeaderExtractor<'a> {
     fn get(&self, key: &str) -> Option<&str> {
-        self.inner.get(key).and_then(|s| s.to_str().ok())
+        self.inner.get(key).and_then(|s| s.iter().next().map(|s| s.as_str()))
     }
 
     fn keys(&self) -> Vec<&str> {
-        self.inner.keys().map(|s| s.as_str()).collect()
+        self.keys.iter().map(|k| k.as_str()).collect()
```

Comment:
Can't you do this by doing something like `self.inner.keys.iter().map(|k| String::from_utf8_lossy(k.as_ref()).as_str()).collect()`. 

If you already tried this and it doesn't work, we should PR an `as_str` function into async_nats. Copying all of the header keys as a `String` is really heavyweight for an OTEL thing. Another alternative could be to use `str::str::from_utf8` to create a `Vec<&str>` as a backup

Reply:
The &str has to be owned by self for it to be in the return value and it's not if String::from_utf8_lossy is used to create the string inside the function.
Agree we should PR async-nats for this feature since it's easy to do there as it's already stored as Strings.